rasterio/_err.pyx

```diff
@@ -205,11 +215,40 @@ cdef OGRErr exc_wrap_ogrerr(OGRErr err) except -1:
         raise CPLE_BaseError(3, err, "OGR Error code {}".format(err))
 
 
+@contextlib.contextmanager
+def stack_errors():
+    """TODO: better name."""
+    # Set up.
+    CPLErrorReset()
+    error_stack = []
+
+    # chaining_error_handler (better name a TODO) records GDAL errors
+    # in the order they occur and converts to exceptions.
+    CPLPushErrorHandlerEx(chaining_error_handler, <void *>error_stack)
+
+    # Run code in the `with` block.
+    yield
+
+    # Link errors via __cause__.
+    for error, cause in zip_longest(error_stack[::-1], error_stack[::-1][1:]):
+        if error is not None and cause is not None:
+            error.__cause__ = cause
+
+    # Tear down.
+    CPLPopErrorHandler()
+    CPLErrorReset()
+
+    # Raise the final exception.
+    if error_stack:
+        last = error_stack.pop()
+        if last is not None:
+            raise last
```

Comment:
This context manager is used like 

```python
try:
    with stack_errors():
        GDALOpenDatasetEx(...)
except CPLE_BaseError as cplerr:
    raise RasterioError("oh no") from cplerr
```

Reply:
Our existing `exc_wrap_pointer` can't do the same thing. The root error is lost before GDALOpenDatasetEx returns.